lib/date_time/date_time_core.c

```diff
@@ -53,39 +56,46 @@ static void date_time_core_notify_event(enum date_time_evt_type time_source)
 	}
 }
 
-static void date_time_core_schedule_update(bool check_pending)
+static void date_time_core_schedule_update(bool fast)
 {
-	/* (Re-)schedule time update work
-	 * if periodic updates are requested in the configuration.
-	 */
-	if (CONFIG_DATE_TIME_UPDATE_INTERVAL_SECONDS > 0) {
-		/* Don't reschedule time update work in some cases,
-		 * e.g. if time is not obtained and the work is already pending.
-		 */
-		if (check_pending && k_work_delayable_is_pending(&date_time_update_work)) {
-			return;
-		}
```

Comment:
I'll comment here about the pending check bug fix:

> From @glarsennordic:
> Also did you see the bugfix commit that removes the `update-in-progress` logic? I did that without consulting anyone on the modem lib team, so I would definitely like to get some direct feedback on that tweak to make sure I am not making a mistake. It certainly works (and without that tweak, the builtin periodic updates do seem to be broken), but it seems worthwhile to have someone more familiar with this area confirm my understanding.

Yep, saw this one. Just did some experiments and seems this is indeed broken. And to clarify to readers. Updates are not always broken but a new update is not scheduled if time is not obtained. When time is obtained, next update has been scheduled correctly so in most cases previous NCS releases are luckily working.
I believe this issue has been introduced when `date_time_update_work_fn` functionality was made from a thread to a work function in a work queue. It's still a bug but this would make sense how it's gone into the code via PR #14108.

It seems this change might introduce another bug/flaw, which `bool checking` parameter has handled even though `k_work_delayable_is_pending` has always returned true. With this change, we will move the time update forward if `date_time_update_async` is called within `CONFIG_DATE_TIME_TOO_OLD_SECONDS`, and we go to these lines of `date_time_update_work_fn`:
```
		LOG_DBG("Using previously obtained time");
		date_time_core_schedule_update(true);
```
While this is not blocking time updates, it'll delay it.

Reply:
> It seems this change might introduce another bug/flaw. With this change, we will move the time update forward if date_time_update_async is called within CONFIG_DATE_TIME_TOO_OLD_SECONDS, and we go to these lines of date_time_update_work_fn:

Oh, interesting. I thought that was the intended behavior anyways, though? Like, I thought it was very much intentional that triggering a time update before `CONFIG_DATE_TIME_TOO_OLD_SECONDS` has elapsed since the previous update would result in `Using previously obtained time`.

If that is not the case, what is the intended behavior in that scenario? Or very possibly I am misunderstanding your description of the possible new bug entirely

Reply:
I'll explain this again with default example values:
`CONFIG_DATE_TIME_UPDATE_INTERVAL_SECONDS=14400` (4h)
`CONFIG_DATE_TIME_TOO_OLD_SECONDS=3600` (1h)

Here is how it goes without this PR:
- `00h00min:` Normal scheduled time update occurs
- `00h50min:` `date_time_update_async()` is called. This goes into `LOG_DBG("Using previously obtained time")` and **won't** reschedule normal time updates.
- `04h00min:` Next normal scheduled time update occurs

And here is how it goes with this PR:
- `00h00min:` Normal scheduled time update occurs
- `00h50min:` `date_time_update_async()` is called. This goes into `LOG_DBG("Using previously obtained time")` and **will** reschedule normal time updates.
- `04h50min:` Next normal scheduled time update occurs

It's another question how big of a problem this is. From functional perspective, I'd say it's non-issue with default values of 4h / 1h. If the values would be 24h / 6h, it becomes potentially a bit worse and 48h / 12h even more.
Probably the worst thing in the change is that when a customer/developer is looking at a log, they might be seeing weird date_time update intervals which they don't necessarily understand from the logs and they do not match with the configuration.

> Oh, interesting. I thought that was the intended behavior anyways, though? Like, I thought it was very much intentional that triggering a time update before `CONFIG_DATE_TIME_TOO_OLD_SECONDS` has elapsed since the previous update would result in `Using previously obtained time`.

Yes, what you say is the designed behavior. However, it's not intentional that the next scheduled time update is rescheduled forward because we haven't really received new time while utilizing the previously received time.

Reply:
I see,

So if there was already a scheduled `normal` update, you don't want `date_time_update_async` to cause that to be rescheduled to later. You want the original `normal` update time to be preserved.

That is pretty easily accomplished with a mutex in a third commit. I'll add it.

Reply:
Looking a bit closer, there is an even simpler solution: We can just use `k_work_schedule` instead of `k_work_reschedule` -- It should behave in exactly the fashion you are desiring.

Reply:
I also looked at `k_work_schedule` yesterday but I think you'd have a bug with retries in that case if it would keep the original scheduling. I.e., when you get `date_time_update_async` call and all methods fail, a retry wouldn't be scheduled after one minute but it would happen with original 4h interval.

Reply:
Yep. In fact, not even the 4h interval reschedule would occur, unfortunately. So I have switched to the atomic idea.